lib/mongo/socket.rb

```diff
@@ -169,11 +169,57 @@ def eof?
     private
 
     def read_from_socket(length)
-      data = String.new
+      # Just in case
+      if length == 0
+        return ''.force_encoding('BINARY')
+      end
+
       deadline = (Time.now + timeout) if timeout
+
+      # We want to have a fixed and reasonably small size buffer for reads
+      # because, for example, OpenSSL reads in 16 kb chunks max.
+      # Having a 16 mb buffer means there will be 1000 reads each allocating
+      # 16 mb of memory and using 16 kb of it.
+      buf_size = read_buffer_size
+      data = nil
+
+      # If we want to read less than the buffer size, just allocate the
+      # memory that is necessary
+      if length < buf_size
+        buf_size = length
+      end
+
+      # The binary encoding is important, otherwise ruby performs encoding
+      # conversions of some sort during the write into the buffer which
+      # kills performance
+      buf = ('x'*buf_size).force_encoding('BINARY')
```

Comment:
You can use `String.new(capacity=buf_size, encoding='BINARY')` in Ruby >= 2.4. Should be faster than 'x'*bufsize, see:
https://ruby-doc.org/core-2.4.0/String.html
https://bugs.ruby-lang.org/issues/12024

Reply:
Unfortunately we support back to 1.9, hence needing to do this manually

Reply:
That being said, it might be worth doing a check here and falling back to manual allocations given that the purpose of these changes are to optimize

Reply:
We do a similar thing here due to the lack of string normalization code in older Ruby versions: https://github.com/mongodb/mongo-ruby-driver/blob/d2090d73dddff8ad1eef950bcbee6781377918ef/lib/mongo/auth/stringprep.rb#L98-L105

Reply:
Thanks for digging that out Shane. It makes a measurable difference for non-SSL/Unix sockets with their 64k buffers, for SSL sockets with 16k buffers the difference is minimal.